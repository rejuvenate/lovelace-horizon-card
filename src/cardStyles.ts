import { css } from 'lit'

export default css`
  .horizon-card {
    --horizon-card-primary: var(--primary-text-color, #000000);
    --horizon-card-secondary: var(--secondary-text-color, #828282);
    --horizon-card-accent: #d7d7d7;

    --horizon-card-lines: var(--horizon-card-accent);
    --horizon-card-field-name-color: var(--horizon-card-secondary);
    --horizon-card-field-value-color: var(--horizon-card-primary);

    --horizon-card-stop-invisible: rgb(0,0,0,0);
    --horizon-card-stop-sun-color: #f9d05e;
    --horizon-card-stop-dawn-color: #393b78;
    --horizon-card-stop-day-color: #8ebeeb;
    --horizon-card-stop-dusk-color: #393b78;

    padding: 0.5rem;
    font-size: 1.3rem;
    font-family: var(--primary-font-family);
  }

  .horizon-card.horizon-card-dark {
    --horizon-card-primary: #ffffff;
    --horizon-card-secondary: #828282;
    --horizon-card-accent: #464646;
  }

  .horizon-card-field-row {
    display: flex;
    justify-content: space-around;
    margin: 1rem 1.5rem 0 1.5rem;
  }

  .horizon-card-text-container {
    display: flex;
    flex-direction: column;
    align-items: center;
  }

  .horizon-card-field-name {
    color: var(--horizon-card-field-name-color);
  }

  .horizon-card-field-value {
    color: var(--horizon-card-field-value-color);
  }

  .horizon-card-header {
    display: flex;
    justify-content: space-between;
    margin: 0 4rem 0 4rem;
  }

  .horizon-card-header .horizon-card-field-value {
    font-size: 1.85rem;
  }

  .horizon-card-title {
    margin: 0.5rem 0.5rem 3rem 3rem;
    font-size: 2rem;
    font-weight: 500;
    color: var(--horizon-card-primary);
  }

  .horizon-card-graph {
    shape-rendering="geometricPrecision";
    margin: 2rem 0 2rem 0;
  }

  .horizon-card-graph .sunInitialStop {
    stop-color: var(--horizon-card-stop-sun-color);
  }

  .horizon-card-graph .sunMiddleStop {
    stop-color: var(--horizon-card-stop-sun-color);
  }

  .horizon-card-graph .sunEndStop {
    stop-color: var(--horizon-card-stop-invisible);
  }

  .horizon-card-graph .dawnInitialStop {
    stop-color: var(--horizon-card-stop-dawn-color);
  }

  .horizon-card-graph .dawnMiddleStop {
    stop-color: var(--horizon-card-stop-dawn-color);
  }

  .horizon-card-graph .dawnEndStop {
    stop-color: var(--horizon-card-stop-invisible);
  }

  .horizon-card-graph .dayInitialStop {
    stop-color: var(--horizon-card-stop-day-color);
  }

  .horizon-card-graph .dayMiddleStop {
    stop-color: var(--horizon-card-stop-day-color);
  }

  .horizon-card-graph .dayEndStop {
    stop-color: var(--horizon-card-stop-invisible);
  }

  .horizon-card-graph .duskInitialStop {
    stop-color: var(--horizon-card-stop-dusk-color);
  }

  .horizon-card-graph .duskMiddleStop {
    stop-color: var(--horizon-card-stop-dusk-color);
  }

  .horizon-card-graph .duskEndStop {
    stop-color: var(--horizon-card-stop-invisible);
  }

  .card-config ul {
    list-style: none;
    padding: 0 0 0 1.5rem;
  }

  .card-config li {
    padding: 0.5rem 0;
  }
`
