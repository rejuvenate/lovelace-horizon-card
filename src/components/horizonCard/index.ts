export * from './HorizonCard'
export * from './HorizonCardContent'
export * from './HorizonCardFooter'
export * from './HorizonCardGraph'
export * from './HorizonCardHeader'
