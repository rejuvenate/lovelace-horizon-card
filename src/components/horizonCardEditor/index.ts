export * from './HorizonCardEditor'
export * from './HorizonCardEditorContent'
