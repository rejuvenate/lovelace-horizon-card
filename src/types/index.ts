import { LovelaceCardConfig } from 'custom-card-helpers'

import { I18N } from '../utils/I18N'

export type THorizonCardFields = {
  sunrise?: boolean
  sunset?: boolean

  dawn?: boolean
  noon?: boolean
  dusk?: boolean

  azimuth?: boolean
  elevation?: boolean
}

export interface IHorizonCardConfig extends LovelaceCardConfig {
  i18n?: I18N
  darkMode?: boolean
  language?: string

  title?: string,
  component?: string
  use12hourClock?: boolean

  fields?: THorizonCardFields
}

export interface IConfigChangedEvent <T> extends CustomEvent<T> {
  target: CustomEvent['target'] & {
    configValue: string
    selected?: string
    checked?: boolean
  }
}

export type TSunInfo = {
  sunrise: number,
  sunset: number

  dawnProgressPercent: number
  dayProgressPercent: number
  duskProgressPercent: number

  sunAboveHorizon: boolean
  sunPercentOverHorizon: number
  sunPosition: {
    x: number
    y: number
  }
}

export enum EHorizonCardErrors {
  SunIntegrationNotFound = 'SunIntegrationNotFound'
}

export type THorizonCardTimes = {
  dawn?: Date
  dusk?: Date
  noon: Date
  sunrise?: Date
  sunset?: Date
}

export type THorizonCardData = {
  azimuth: number
  elevation: number

  sunInfo: TSunInfo

  times: THorizonCardTimes

  error?: EHorizonCardErrors
}

export enum EHorizonCardI18NKeys {
  Azimuth = 'azimuth',
  Dawn = 'dawn',
  Dusk = 'dusk',
  Elevation = 'elevation',
  Noon = 'noon',
  Sunrise = 'sunrise',
  Sunset = 'sunset'
}

export type THorizonCardI18N = Record<string, unknown>

export type THorizonCardI18NErrorKeys = {
  [key in EHorizonCardErrors]: string
}

export type THorizonCardI18NKeys = { [key in EHorizonCardI18NKeys ]: string } | { errors: THorizonCardI18NErrorKeys }
