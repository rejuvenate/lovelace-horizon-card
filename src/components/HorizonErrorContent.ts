import { html, TemplateResult } from 'lit'

import { EHorizonCardErrors, IHorizonCardConfig } from '../types'
import { I18N } from '../utils/I18N'

export class HorizonErrorContent {
  private i18n: I18N
  private error: EHorizonCardErrors

  constructor (config: IHorizonCardConfig, error: EHorizonCardErrors) {
    this.i18n = config.i18n!
    this.error = error
  }

  public render (): TemplateResult {
    const errorMessage = this.i18n.tr(`errors.${this.error}`)
    // eslint-disable-next-line no-console
    console.error(errorMessage)

    return html`
      <div class="horizon-card-error">
        ${ errorMessage }
      </div>
    `
  }
}
