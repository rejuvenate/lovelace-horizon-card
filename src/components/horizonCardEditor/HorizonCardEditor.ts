import { HomeAssistant, LovelaceCardEditor } from 'custom-card-helpers'
import { CSSResult, LitElement, TemplateResult } from 'lit'
import { customElement, property } from 'lit-element'

import cardStyles from '../../cardStyles'
import { EHorizonCardI18NKeys, IHorizonCardConfig } from '../../types'
import { HorizonCardEditorContent, THorizonCardEditorContentEvents } from './HorizonCardEditorContent'

@customElement('horizon-card-editor')
export class HorizonCardEditor extends LitElement implements LovelaceCardEditor {
  static readonly cardType = 'horizon-card-editor'
  private static readonly CONFIG_CHANGED_EVENT = 'config-changed'

  @property({ type: Object }) hass!: HomeAssistant
  @property() private config!: IHorizonCardConfig

  static get styles (): CSSResult {
    return cardStyles
  }

  public setConfig (config: IHorizonCardConfig): void {
    this.config = config
  }

  public configChanged (event: THorizonCardEditorContentEvents['configChanged']): void {
    const property = event.target?.configValue
    const value = event.detail?.value ?? event.target?.selected ?? event.target?.checked

    const newConfig = { ...this.config, [property]: value }

    // Handles default or empty values by deleting the config property
    if (value === 'default' || value === undefined || value === '') {
      delete newConfig[property]
    }

    // Handles boolean values
    if (value === 'true' || value === 'false') {
      newConfig[property] = value === 'true'
    }

    // Handles fields config
    if (Object.values(EHorizonCardI18NKeys).includes(property as EHorizonCardI18NKeys)) {
      delete newConfig[property]
      newConfig.fields = {
        ...newConfig.fields,
        [property]: value
      }
    }

    const customEvent = new CustomEvent(HorizonCardEditor.CONFIG_CHANGED_EVENT, {
      bubbles: true,
      composed: true,
      detail: { config: newConfig }
    })

    this.dispatchEvent(customEvent)
  }

  protected render (): TemplateResult {
    const content = new HorizonCardEditorContent(this.config!)
    content.on('configChanged', (event) => this.configChanged(event))
    return content.render()
  }
}
