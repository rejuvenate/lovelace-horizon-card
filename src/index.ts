import './components/horizonCard'
import './components/horizonCardEditor'

