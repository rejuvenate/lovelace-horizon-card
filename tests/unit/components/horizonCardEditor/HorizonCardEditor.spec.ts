import { HomeAssistant } from 'custom-card-helpers'
import { css, CSSResult } from 'lit'

import { HorizonCardEditor, THorizonCardEditorContentEvents } from '../../../../src/components/horizonCardEditor'
import { IHorizonCardConfig } from '../../../../src/types'
import { CustomSnapshotSerializer, TemplateResultTestHelper } from '../../../helpers/TestHelpers'
import { HorizonCardEditorContent } from '../../../mocks/HorizonCardEditorContent'

jest.mock('../../../../src/components/HorizonErrorContent', () => require('../../../mocks/HorizonErrorContent'))
jest.mock('../../../../src/utils/HelperFunctions', () => require('../../../mocks/HelperFunctions'))
jest.mock('../../../../src/utils/I18N', () => require('../../../mocks/I18N'))
jest.mock('../../../../src/cardStyles', () => css``)
jest.mock('../../../../src/components/horizonCardEditor/HorizonCardEditorContent.ts', () => require('../../../mocks/HorizonCardEditorContent'))

expect.addSnapshotSerializer(new CustomSnapshotSerializer())

describe('HorizonCardEditor', () => {
  let horizonCardEditor: HorizonCardEditor

  beforeEach(() => {
    HorizonCardEditorContent.onMock = jest.fn()
    horizonCardEditor = new HorizonCardEditor()
  })

  describe('hass', () => {
    it('updates hass property with provided value', () => {
      const expectedValue = { language: 'test' } as HomeAssistant
      horizonCardEditor.hass = expectedValue
      expect(horizonCardEditor.hass).toEqual(expectedValue)
    })
  })

  describe('setConfig', () => {
    it('updates config property with provided value', () => {
      const expectedValue = { language: 'test' } as IHorizonCardConfig
      horizonCardEditor.setConfig(expectedValue)
      expect(horizonCardEditor['config']).toEqual(expectedValue)
    })
  })

  describe('configChanged', () => {
    let dispatchEventMock
    beforeEach(() => {
      dispatchEventMock = jest.fn()
      horizonCardEditor.dispatchEvent = dispatchEventMock
    })

    it('retrieves correctly the value from input', () => {
      const config = { target: { configValue: 'test' }, detail: { value: 'test' } } as THorizonCardEditorContentEvents['configChanged']
      horizonCardEditor.configChanged(config)

      const customEvent = dispatchEventMock.mock.calls[0][0] as CustomEvent
      expect(customEvent.detail.config).toEqual(expect.objectContaining({
        test: 'test'
      }))
    })

    it('retrieves correctly the value from select', () => {
      const config = { target: { configValue: 'test', selected: 'true' } } as THorizonCardEditorContentEvents['configChanged']
      horizonCardEditor.configChanged(config)

      const customEvent = dispatchEventMock.mock.calls[0][0] as CustomEvent
      expect(customEvent.detail.config).toEqual(expect.objectContaining({
        test: true
      }))
    })

    it('retrieves correctly the value from checkbox', () => {
      const config = { target: { configValue: 'test', checked: true } } as THorizonCardEditorContentEvents['configChanged']
      horizonCardEditor.configChanged(config)

      const customEvent = dispatchEventMock.mock.calls[0][0] as CustomEvent
      expect(customEvent.detail.config).toEqual(expect.objectContaining({
        test: true
      }))
    })

    it('deletes the config property when it the value is default', () => {
      const config = { target: { configValue: 'test' }, detail: { value: 'default' } } as THorizonCardEditorContentEvents['configChanged']
      horizonCardEditor.configChanged(config)

      const customEvent = dispatchEventMock.mock.calls[0][0] as CustomEvent
      expect(customEvent.detail.config.test).toBeUndefined()
    })

    it('deletes the config property when it the value is undefined', () => {
      const config = { target: { configValue: 'test' } } as THorizonCardEditorContentEvents['configChanged']
      horizonCardEditor.configChanged(config)

      const customEvent = dispatchEventMock.mock.calls[0][0] as CustomEvent
      expect(customEvent.detail.config.test).toBeUndefined()
    })

    it('deletes the config property when it the value is an empty string', () => {
      const config = { target: { configValue: 'test' }, detail: { value: '' } } as THorizonCardEditorContentEvents['configChanged']
      horizonCardEditor.configChanged(config)

      const customEvent = dispatchEventMock.mock.calls[0][0] as CustomEvent
      expect(customEvent.detail.config.test).toBeUndefined()
    })

    it('transform "true"/"false" to boolean', () => {
      const config = { target: { configValue: 'test' }, detail: { value: 'false' } } as THorizonCardEditorContentEvents['configChanged']
      horizonCardEditor.configChanged(config)

      const customEvent = dispatchEventMock.mock.calls[0][0] as CustomEvent
      expect(customEvent.detail.config).toEqual(expect.objectContaining({
        test: false
      }))
    })

    it('handles field configuration correctly', () => {
      const config = { target: { configValue: 'dawn' }, detail: { value: true } } as THorizonCardEditorContentEvents['configChanged']
      horizonCardEditor.configChanged(config)

      const customEvent = dispatchEventMock.mock.calls[0][0] as CustomEvent
      expect(customEvent.detail.config).toEqual(expect.objectContaining({
        fields: {
          dawn: true
        }
      }))
    })

    it('dispatches an event', () => {
      const config = { target: { configValue: 'test' }, detail: { value: 'test' } } as THorizonCardEditorContentEvents['configChanged']
      horizonCardEditor.configChanged(config)

      expect(horizonCardEditor.dispatchEvent).toHaveBeenCalledTimes(1)
    })
  })

  describe('render', () => {
    it('starts to listen then configChanged event from HorizonCardEditorContent', () => {
      horizonCardEditor['render']()
      expect(HorizonCardEditorContent.onMock).toHaveBeenCalledWith('configChanged', expect.any(Function))
    })

    it('renders the horizon card editor content render result', async () => {
      const element = window.document.createElement('test-element') as TemplateResultTestHelper<typeof horizonCardEditor['render']>
      element.templateResultFunction = () => horizonCardEditor['render']()
      window.document.body.appendChild(element)
      await element.updateComplete

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot()
    })
  })

  describe('get styles', () => {
    it('returns a CSSResult', () => {
      expect(HorizonCardEditor.styles).toBeInstanceOf(CSSResult)
    })
  })
})
