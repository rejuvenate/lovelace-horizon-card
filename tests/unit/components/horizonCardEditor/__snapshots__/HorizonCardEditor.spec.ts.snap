// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HorizonCardEditor render renders the horizon card editor content render result 1`] = `
<!---->
      <div class="card-config">
        HORIZON CARD EDITOR CONTENT
      </div>
    
`;
